streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -378,42 +320,88 @@ public void prepareCommit() {
         }
+
+        return committableOffsetsAndMetadata();
     }
 
-    @Override
-    public void postCommit() {
+    private Map<TopicPartition, OffsetAndMetadata> committableOffsetsAndMetadata() {
+        final Map<TopicPartition, OffsetAndMetadata> committableOffsets;
+
         switch (state()) {
-            case RUNNING:
-                commitNeeded = false;
-                commitRequested = false;
+            case CREATED:
+            case RESTORING:
+                committableOffsets = Collections.emptyMap();
 
-                if (!eosEnabled) {
-                    stateMgr.checkpoint(checkpointableOffsets());
-                }
+                break;
 
-                log.debug("Committed");
+            case RUNNING:
+            case SUSPENDED:
+                final Map<TopicPartition, Long> partitionTimes = extractPartitionTimes();
+
+                committableOffsets = new HashMap<>(consumedOffsets.size());
+                for (final Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
+                    final TopicPartition partition = entry.getKey();
+                    Long offset = partitionGroup.headRecordOffset(partition);
+                    if (offset == null) {
+                        try {
+                            offset = mainConsumer.position(partition);
+                        } catch (final TimeoutException error) {
+                            // the `consumer.position()` call should never block, because we know that we did process data
+                            // for the requested partition and thus the consumer should have a valid local position
+                            // that it can return immediately
+
+                            // hence, a `TimeoutException` indicates a bug and thus we rethrow it as fatal `IllegalStateException`
+                            throw new IllegalStateException(error);
+                        } catch (final KafkaException fatal) {
+                            throw new StreamsException(fatal);
+                        }
+                    }
+                    final long partitionTime = partitionTimes.get(partition);
+                    committableOffsets.put(partition, new OffsetAndMetadata(offset, encodeTimestamp(partitionTime)));
+                }
 
                 break;
 
+            case CLOSED:
+                throw new IllegalStateException("Illegal state " + state() + " while getting commitable offsets for active task " + id);
+
+            default:
+                throw new IllegalStateException("Unknown state " + state() + " while post committing active task " + id);
+        }
+
+        return committableOffsets;
+    }
+
+    @Override
+    public void postCommit() {
+        switch (state()) {
             case RESTORING:
-                commitNeeded = false;
+            case RUNNING:
+            case SUSPENDED:
```

Comment:
What about 1) move the line 387/388 out of the switch, also line 400 after the switch block, 2) and then make these three states separate branches, so that we can avoid a mix of switch / if-else.